docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,505 @@
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+      - [Test Plan [optional]](#test-plan-optional)
+      - [Graduation Criteria [optional]](#graduation-criteria-optional)
+      - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+### ClusterClass
+A collection of templates that define a topology (control plane and machine deployments) to be used in one or more Managed Clusters.
+
+### Managed Cluster topology
+A Managed Cluster is a Cluster that provides a single control point to manage it's own topology; the managed cluster topology is defined by a ClusterClass.
+
+### Worker class
+A worker class is a collection of templates that define a set of worker nodes in the cluster.
+
+
+## Summary
+
+This proposal introduces a new ClusterClass object which will be used to provide easy stamping of clusters of similar shapes. It serves as a collection of template resources which are used to generate one or more clusters of the same flavor.
+
+We're enhancing the Cluster CRD and controller to use a ClusterClass resource to provision the underlying objects that compose a cluster. Additionally, the Cluster provides a single control point to manage the Kubernetes version, worker pools, labels, replicas, and so on.
+
+## Motivation
+
+Currently, Cluster API does not expose a native way to provision multiple clusters of the same configuration. The ClusterClass object is supposed to act as a collection of template references which can be used to create managed topologies.
+
+Today, the Cluster object is a logical grouping of components which describe an underlying cluster. The user experience to create a cluster requires the user to create a bunch of underlying resources such as KCP (control plane provider), MachineDeployments, and infrastructure or bootstrap templates for those resources which logically end up representing the cluster. Since the cluster configuration is spread around multiple components, upgrading the cluster version is hard as it requires changes to different fields in different resources to perform an upgrade.
+
+This method of provisioning the cluster would act as a single control point for the entire cluster. Scaling the nodes, adding/removing node pools and upgrading cluster kubernetes versions would be achievable by editing the managed topology. This would facilitate the maintenance of existing clusters as well as ease the creation of newer clusters.
+
+### Goals
+
+- Create the new ClusterClass CRD which can serve as a collection of templates to create clusters.
+- Extend the Cluster object to use ClusterClass for creating  managed cluster topologies.
+- Enhance the Cluster object to act as a single point of control for the managed topology.
+- Extend the Cluster controller to create/update/delete managed topologies.
```

Comment:
This sounds the same than the future work below:
`Introduce mechanisms for allowing mutation of a ClusterClass, and the continuous reconciliation of the Cluster managed resources.`

So may be update is not part of the goal?

Reply:
I think what's meant here is that the `ClusterSpec` will be extended by a `managed` field, which contains the reference to a ClusterClass and config like the amount of nodes.

In order for ClusterClass to  be usable, the Cluster controller needs to reconcile that new field and take care of creating the resources as specified in the ClusterClass.

Reply:
Like @schrej mentioned above, the update is to handle the config updates to `spec.Managed` field.